damage.py

```diff
@@ -1,9 +1,158 @@
 # TODO: it seems that hp are ceiled while damage floored. (however it seems that the ad is rounded in the stat section ingame)
 
-def damage_ad_armor(attack_ad: float, defense_armor: float):
+def physical_damage_after_positive_armor(pre_mitigation_damage: float, defense_armor: float):
     """
-    Calculates the output damage of an auto attack from a champion with X amount of 
-    attack damage on a champion with Y amount of armor.
+    Calculates the output damage if X amount of pre-mitigation physical damage is dealt to a champion with Y amount of
+    armor (positive)
     """
-    return attack_ad * 100 / (100 + defense_armor)
+    return pre_mitigation_damage * 100 / (100 + defense_armor)
 
+
+def physical_damage_after_negative_armor(pre_mitigation_damage: float, defense_armor: float):
+    """
+    Calculates the output damage if X amount of pre-mitigation physical damage is dealt to a champion with negative
+    armor.
+    """
+    return pre_mitigation_damage * (2 - 100 / (100 - defense_armor))
+
+
+def damage_normal_auto_attack_no_crit(attack_base_ad: float, attack_bonus_ad: float, attack_lethality: float,
+                                      attack_level: int, attack_armor_pen: float, attack_bonus_armor_pen: float,
+                                      defense_base_armor: float, defense_bonus_armor: float):
+    """
+    Calculates the output damage of an auto attack that doesn't crit
+    The base armor and bonus armor of the champion being attacked should already take into account the flat or
+    percentage armor reduction resulting from spells like Garen E, Trundle R, Olaf Q, Corki E, or items like Black Cleaver
+    """
+    pre_mitigation_damage = attack_base_ad + attack_bonus_ad
+    defense_armor = defense_base_armor + defense_bonus_armor
+    if defense_armor < 0:
+        return physical_damage_after_negative_armor(pre_mitigation_damage, defense_armor)
+    else:
+        attack_flat_armor_pen = attack_lethality * (0.6 + 0.4 * attack_level / 18)
+        armor_eq = defense_base_armor * (1 - attack_armor_pen) + defense_bonus_armor * (1 - attack_armor_pen) * (
```

Comment:
percentage armor pen and armor reduction should be represented directly as the multiplicative factor (post `1-x`) so that it can be stacked multiplicatively.

If you have two items with 10% armor pen, it's not immediate what the result X% armor pen is, whereas if you view them as x0.9 armor pen then you can just multiply them together. So 10% armor pen should be represented as 0.9 to begin with.

Reply:
bien vu !